ipld/amt/src/iter.rs

```diff
@@ -0,0 +1,364 @@
+// Copyright 2021-2023 Protocol Labs
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+use crate::node::CollapsedNode;
+use crate::node::{Link, Node};
+use crate::MAX_INDEX;
+use crate::{nodes_for_height, Error};
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::ser::Serialize;
+use fvm_ipld_encoding::CborStore;
+use serde::de::DeserializeOwned;
+use std::marker::PhantomData;
+
+impl<V, BS, Ver> crate::AmtImpl<V, BS, Ver>
+where
+    V: DeserializeOwned + Serialize,
+    BS: Blockstore,
+    Ver: crate::root::version::Version,
+{
+    /// Iterate over the AMT. Alternatively, you can directly iterate over the AMT without calling
+    /// this method:
+    ///
+    /// ```rust
+    /// use fvm_ipld_amt::Amt;
+    /// use fvm_ipld_blockstore::MemoryBlockstore;
+    ///
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// let mut amt = Amt::new(store);
+    /// let kvs: Vec<u64> = (0..=5).collect();
+    /// kvs
+    ///     .iter()
+    ///     .map(|k| amt.set(u64::try_from(*k).unwrap(), k.to_string()))
+    ///     .collect::<Vec<_>>();
+    ///
+    /// for kv in &amt {
+    ///     let (k, v) = kv?;
+    ///     println!("{k:?}: {v:?}");
+    /// }
+    ///
+    /// # anyhow::Ok(())
+    /// ```
+    pub fn iter(&self) -> Iter<'_, V, &BS, Ver> {
+        Iter {
+            stack: vec![IterStack {
+                node: &self.root.node,
+                idx: 0,
+                height: self.root.height,
+            }],
+            blockstore: &self.block_store,
+            bit_width: self.bit_width(),
+            ver: PhantomData,
+            key: 0,
+        }
+    }
+}
+
+impl<'a, V, BS, Ver> IntoIterator for &'a crate::AmtImpl<V, BS, Ver>
+where
+    V: Serialize + DeserializeOwned,
+    Ver: crate::root::version::Version,
+    BS: Blockstore,
+{
+    type IntoIter = Iter<'a, V, &'a BS, Ver>;
+    type Item = Result<(u64, &'a V), crate::Error>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+pub struct Iter<'a, V, BS, Ver> {
+    stack: Vec<IterStack<'a, V>>,
+    blockstore: BS,
+    bit_width: u32,
+    ver: PhantomData<Ver>,
+    key: u64,
+}
+
+pub struct IterStack<'a, V> {
+    pub(crate) node: &'a Node<V>,
+    pub(crate) idx: usize,
+    pub(crate) height: u32,
+}
+
+impl<'a, V, BS, Ver> Iterator for Iter<'a, V, BS, Ver>
+where
+    BS: Blockstore,
+    V: Serialize + DeserializeOwned,
+{
+    type Item = Result<(u64, &'a V), crate::Error>;
+    fn next(&mut self) -> Option<Self::Item> {
+        let root_height = self.stack[0].height;
```

Comment:
Won't this panic if we call `next` after exhausting the iterator? Technically that's allowed by spec, but it's not very nice.

I'd either:

1. Store the height in `Iter` _only_, always calculating the current height as `self.height - self.stack.len()`.
2. Compute the next height as `prev_stack.height - 1`.